File: CPP/No115.cc
/**
 * Created by Xiaozhong on 2020/8/5.
 * Copyright (c) 2020/8/5 Xiaozhong. All rights reserved.
 */

#include "iostream"
#include "vector"
#include "string"

using namespace std;

class Solution {
public:
    int numDistinct(string s, string t) {
        long dp[t.size() + 1][s.size() + 1];
        for (int i = 0; i < s.size() + 1; i++) dp[0][i] = 1;
        for (int i = 1; i < t.size() + 1; i++) {
            for (int j = 1; j < s.size() + 1; j++) {
                if (s[j - 1] == t[i - 1])  dp[i][j] = dp[i - 1][j - 1] + dp[i][j - 1];
                else dp[i][j] = dp[i][j - 1];
            }
        }
        return dp[t.length()][s.length()];
    }
};

File: Java/src/No115.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

public class No115 {
    public int numDistinct(String s, String t) {
        int n = s.length(), m = t.length();
        if (n < m) return 0;
        int[][] dp = new int[n + 1][m + 1];
        for (int i = 0; i <= n; ++i) dp[i][m] = 1;
        for (int sIndex = n - 1; sIndex >= 0; --sIndex) {
            for (int tIndex = m - 1; tIndex >= 0; --tIndex) {
                char cs = s.charAt(sIndex);
                char ct = t.charAt(tIndex);
                if (cs == ct) dp[sIndex][tIndex] = dp[sIndex + 1][tIndex] + dp[sIndex + 1][tIndex + 1];
                else dp[sIndex][tIndex] = dp[sIndex + 1][tIndex];
            }
        }
        return dp[0][0];
    }

    public static void main(String[] args) {
        No115 no115 = new No115();
        System.out.println(no115.numDistinct("rabbbit", "rabbit"));
    }
}
